core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/PromiseManager.scala

```diff
@@ -0,0 +1,234 @@
+            unCompletedGroupPromises.remove(i)
+          }
+        }
+
+      // handle root synchronous promise
+      case PromiseInvocation(
+            ctx: RootPromiseContext,
+            call: PromiseBody[_] with SynchronizedInvocation
+          ) =>
+        if (syncPromiseRoot == null) {
+          // if there is no other executing synchronous promise,
+          // execute this one
+          registerSyncPromise(ctx, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // handle synchronous promise created by other promise
+      case PromiseInvocation(
+            ctx: ChildPromiseContext,
+            call: PromiseBody[_] with SynchronizedInvocation
+          ) =>
+        if (syncPromiseRoot == null || ctx.root == syncPromiseRoot) {
+          // if there is no other executing synchronous promise,
+          // or this promise has same root context (chain of promises),
+          // execute this one.
+          registerSyncPromise(ctx.root, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // trivial case
+      case p: PromiseInvocation =>
+        // set context
+        promiseContext = p.context
+        try {
+          // execute it
+          promiseHandler(p.call)
+        } catch {
+          case e: Throwable =>
+            // if error occurs, return it to the sender.
+            returning(e)
+        }
+    }
+
+    // execute queued sync promises
+    tryInvokeNextSyncPromise()
+  }
+
+  // send a control message to another actor, and keep the handle.
+  def schedule[T](cmd: PromiseBody[T], on: ActorVirtualIdentity): Promise[T] = {
+    val (promise, ctx) = createPromise[T]()
+    controlOutputPort.sendTo(on, PromiseInvocation(ctx, cmd))
+    promise
+  }
+
+  // send a grouped control message to other actors, and keep the handle.
+  def schedule[T: ClassTag](seq: (PromiseBody[T], ActorVirtualIdentity)*): Promise[Seq[T]] = {
+    val promise = WorkflowPromise[Seq[T]](promiseContext)
+    if (seq.isEmpty) {
+      // if the sequence is empty, resolve the promise immediately
+      promise.setValue(Seq.empty)
+    } else {
+      unCompletedGroupPromises.add(
+        GroupedWorkflowPromise[T](promiseID, promiseID + seq.length, promise)
+      )
+      seq.foreach {
+        case (body, virtualIdentity) =>
+          val ctx = mkPromiseContext()
+          promiseID += 1
+          controlOutputPort.sendTo(virtualIdentity, PromiseInvocation(ctx, body))
+      }
+    }
+    promise
+  }
+
+  def returning(value: Any): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(promiseContext.sender, ReturnPayload(promiseContext, value))
+      exitCurrentPromise()
+    }
+  }
+
+  def returning(): Unit = {
```

Comment:
returning is called when a call has B called C, and C has responded and B has received the response. So, B now will open the returnpayload and take out the context. From the map, it should take out the WorkflowPromise that has the callerPromiseContext and then call returning(callerPromiseContext).

Especially if A has called B and D has also called B for some other reason, then the two promisecontexts can overlap and map is the right way to get it.